druid/src/main/java/org/apache/calcite/adapter/druid/DruidQuery.java

```diff
@@ -346,7 +346,28 @@ public DruidTable getDruidTable() {
 
   @Override public RelOptCost computeSelfCost(RelOptPlanner planner,
       RelMetadataQuery mq) {
-    return Util.last(rels).computeSelfCost(planner, mq).multiplyBy(.1);
+    return Util.last(rels)
+               .computeSelfCost(planner, mq)
+               // Cost is directly proportional to number of fields queried.
+               // A plan where all extra columns are pruned will be preferred.
+               .multiplyBy(querySpec.fieldNames.size())
+               .multiplyBy(getQueryTypeCostMultiplier());
```

Comment:
I think this is too sharply linear in the number of fields. When I commit, I'll add a function to soften it, so that 3 or fewer has "normal" cost, and 100 or more has 2x the cost.